contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1051 @@
+        require(loanOffering.positionOwner == margin.getPositionOwner(POSITION_ID));
+        require(loanOffering.lenderFeeToken == address(0));
+        require(loanOffering.takerFeeToken == address(0));
+
+        // CHECK VALUES256
+        require(loanOffering.maximumAmount == MathHelpers.maxUint256());
+        require(loanOffering.minimumAmount == 0);
+        require(loanOffering.minimumHeldToken == 0);
+        require(loanOffering.rates.lenderFee == 0);
+        require(loanOffering.rates.takerFee == 0);
+        require(loanOffering.expirationTimestamp == MathHelpers.maxUint256());
+        require(loanOffering.salt == 0);
+
+        // CHECK VALUES32
+        require(loanOffering.callTimeLimit == margin.getPositionCallTimeLimit(POSITION_ID));
+        require(loanOffering.maxDuration == margin.getPositionMaxDuration(POSITION_ID));
+        require(loanOffering.interestRate == margin.getPositionInterestRate(POSITION_ID));
+        require(loanOffering.interestPeriod == margin.getPositioninterestPeriod(POSITION_ID));
+
+        // CHECK SIGNATURE
+        // no need to require anything about loanOffering.signature
+        */
+
+        return address(this);
+    }
+
+    /**
+     * Called by the Margin contract when anyone transfers ownership of a loan to this contract.
+     * This function initializes this contract and returns this address to indicate to Margin
+     * that it is willing to take ownership of the loan.
+     *
+     * @param  from        (unused)
+     * @param  positionId  Unique ID of the position
+     * @return             This address on success, throw otherwise
+     */
+    function receiveLoanOwnership(
+        address from,
+        bytes32 positionId
+    )
+        external
+        onlyMargin
+        onlyPosition(positionId)
+        returns (address)
+    {
+        MarginCommon.Position memory position = MarginHelper.getPosition(DYDX_MARGIN, POSITION_ID);
+
+        assert(principalTotal == 0);
+        assert(position.principal > 0);
+        assert(position.owedToken == OWED_TOKEN);
+        assert(position.heldToken == HELD_TOKEN);
+
+        // set relevant constants
+        uint256 initialPrincipal = position.principal;
+        principalForBucket[0] = initialPrincipal;
+        principalTotal = initialPrincipal;
+        weightForBucket[0] = weightForBucket[0].add(initialPrincipal);
+        weightForBucketForAccount[0][from] =
+            weightForBucketForAccount[0][from].add(initialPrincipal);
+
+        return address(this);
+    }
+
+    /**
+     * Called by Margin when additional value is added onto the position this contract
+     * is lending for. Balance is added to the address that loaned the additional tokens.
+     *
+     * @param  payer           Address that loaned the additional tokens
+     * @param  positionId      Unique ID of the position
+     * @param  principalAdded  Amount that was added to the position
+     *  param  lentAmount      (unused)
```

Comment:
not unused (just check all these comments / not sure if you've updated them all yet)